hedera-node/src/main/java/com/hedera/services/records/ConsensusTimeTracker.java

```diff
@@ -0,0 +1,323 @@
+
+	/**
+	 * The max number of records that can be placed before transactions in a
+	 * {@link com.hedera.services.state.logic.StandardProcessLogic#incorporateConsensusTxn} call.
+	 *
+	 * There is a special case. When the system first starts up, the first
+	 * {@link com.hedera.services.state.logic.StandardProcessLogic#incorporateConsensusTxn} call can
+	 * haven an unlimited number of preceding records. This is for migrations.
+	 */
+	public static final long MAX_PRECEDING_RECORDS = 3;
+
+	/**
+	 * The max number of records that can be placed after a transaction in a
+	 * {@link com.hedera.services.state.logic.StandardProcessLogic#incorporateConsensusTxn} call
+	 */
+	public static final long MAX_FOLLOWING_RECORDS = 10;
+
+	private final long nanosBetweenIncorporateCalls;
+
+	private Instant minConsensusTime;
+	private Instant currentTxnMinTime;
+	private Instant currentTxnTime;
+	private Instant currentTxnMaxTime;
+	private Instant maxConsensusTime;
+
+	private long followingRecordsCount;
+
+	private boolean isFirstSinceStartup = true;
+
+	private boolean firstUsed;
+
+	@Inject
+	public ConsensusTimeTracker() {
+		this(DEFAULT_NANOS_PER_INCORPORATE_CALL);
+	}
+
+	@VisibleForTesting
+	ConsensusTimeTracker(final long nanosBetweenIncorporateCalls) {
+		this.nanosBetweenIncorporateCalls = nanosBetweenIncorporateCalls;
+		if (nanosBetweenIncorporateCalls <
+				(MAX_FOLLOWING_RECORDS + MAX_PRECEDING_RECORDS + 10)) {
+			throw new IllegalArgumentException("TransactionHandler.MIN_TRANS_TIMESTAMP_INCR_NANOS is too small!");
+		}
+	}
+
+	@VisibleForTesting
+	ConsensusTimeTracker(ConsensusTimeTracker toCopy) {
+		this();
+		minConsensusTime = toCopy.minConsensusTime;
+		currentTxnMinTime = toCopy.currentTxnMinTime;
+		currentTxnTime = toCopy.currentTxnTime;
+		currentTxnMaxTime = toCopy.currentTxnMaxTime;
+		maxConsensusTime = toCopy.maxConsensusTime;
+		followingRecordsCount = toCopy.followingRecordsCount;
+		firstUsed = toCopy.firstUsed;
+		isFirstSinceStartup = toCopy.isFirstSinceStartup;
+	}
+
+	/**
+	 * @param count the actual number of records that were placed following the consensusTime of the current transaction.
+	 */
+	public void setActualFollowingRecordsCount(final long count) {
+		if (count < 0) {
+			throw new IllegalArgumentException("count must be >= 0, it was " + count);
+		}
+		followingRecordsCount = count;
+	}
+
+	/**
+	 * @return true if firstTransactionTime() cannot be used again in the current
+	 * {@link com.hedera.services.state.logic.StandardProcessLogic#incorporateConsensusTxn} call
+	 */
+	public boolean isFirstUsed() {
+		return firstUsed;
+	}
+
+	/**
+	 * Resets this tracker, should be called at the beginning of
+	 * {@link com.hedera.services.state.logic.StandardProcessLogic#incorporateConsensusTxn}
+	 * @param consensusTime the consensusTime passed to incorporateConsensusTxn
+	 */
+	public void reset(final Instant consensusTime) {
```

Comment:
Would it be possible to implement this class with the `Instant minConsensusTime` field, and then `int`s for `preTimesUsedByCurrent`, `postTimesUsedByCurrent`, etc?

May seem a bit crazy, but when on a microsecond budget, the three extra `Instant` creations here, and one extra creation in `hasMoreTime()`, actually do feel indulgent...

Reply:
I'll see what I can do.

Reply:
This is done.